### PR TITLE
Dev generalized quantitative support feature

### PR DESCRIPTION
**Pull request by ccssu**
- 量化过程耗时：sdxl 跑量化校准信息耗时 4 min(NVIDIA GeForce RTX 3090)
-  因为是在线量化，可依同时支持 diffusers , comfyui, sd_webui

[sd_v1-5:conv_linear_calibrate_info](https://ccssu.github.io/git_pages.io/run_sd_v1-5_deepcache/conv_linear_calibrate_infoUNet2DConditionModel_06f5730368ef1a128ac8e3ae1728b45f2c7236621a050beb40bba85b1ef7b660.html)

 [sdxl:conv_linear_calibrate_info](https://ccssu.github.io/git_pages.io/run_sdxl/conv_linear_calibrate_infoUNet2DConditionModel_34f5bd9d3416cc1703c40491d1fa89fc50081c344f7a9ccf7b64f853a711f342.html)

下图 sd_v1-5:conv_linear_calibrate_info ，横轴 需要量化 layer 的顺序 , 纵轴 归一化 的数据值。
newplot (2)

<details close>
<summary> ### SD_v1-5 </summary>

> compute_density_threshold = 100

- [sd_v1-5:conv_linear_calibrate_info](https://ccssu.github.io/git_pages.io/run_sd_v1-5_deepcache/conv_linear_calibrate_infoUNet2DConditionModel_06f5730368ef1a128ac8e3ae1728b45f2c7236621a050beb40bba85b1ef7b660.html)

- [sd_v1-5:const_cailbrate_info](https://ccssu.github.io/git_pages.io/run_sd_v1-5/CostsQuantizationCalibratorUNet2DConditionModel_06f5730368ef1a128ac8e3ae1728b45f2c7236621a050beb40bba85b1ef7b660.json.html)

| sd_v1-5 | quality_level | SSIM               | Conv-Count | Linear-Count |
| ------- | ------------- | ------------------ | ---------- | ------------ |
| sd_v1-5 | 0             | 0.8165556563837066 | 98         | 184          |
| sd_v1-5 | 1             | 0.8323949474268865 | 96         | 160          |
| sd_v1-5 | 2             | 0.8348702314076454 | 88         | 142          |
| sd_v1-5 | 3             | 0.8662256561375715 | 84         | 139          |
| sd_v1-5 | 4             | 0.8707190288604787 | 62         | 113          |

| sd_v1-5          | PyTorch                                                     | OneDiff                                                       | SSIM   |
| ---------------- | ----------------------------------------------------------- | ------------------------------------------------------------- | ------ |
|                  |  |  | 0.9838 |
| quality_level: 0 |  |        | 0.8166 |
| quality_level: 1 |  |        | 0.8324 |
| quality_level: 2 |  |        | 0.8349 |
| quality_level: 3 |  |        | 0.8662 |
| quality_level: 4 |  |        | 0.8816 |

</details>

<details close>
<summary> ### SD_v1-5 DeepCache </summary>

| sd_v1-5_deepcache | quality_level | SSIM               | Conv-Count | Linear-Count |
| ----------------- | ------------- | ------------------ | ---------- | ------------ |
| sd_v1-5_deepcache | 0             | 0.8517729834085559 | 0          | 184          |
| sd_v1-5_deepcache | 1             | 0.8469390501432797 | 0          | 160          |
| sd_v1-5_deepcache | 2             | 0.8460595799988786 | 0          | 141          |
| sd_v1-5_deepcache | 3             | 0.852704767430985  | 0          | 116          |
| sd_v1-5_deepcache | 4             | 0.8509655270067743 | 0          | 109          |

| sd_v1-5_deepcache | PyTorch                                                               | OneDiff                                                           | SSIM   |
| ----------------- | --------------------------------------------------------------------- | ----------------------------------------------------------------- | ------ |
|                   |  |        | 0.8682 |
| quality_level: 0  |  |  | 0.8518 |
| quality_level: 1  |  |  | 0.8469 |
| quality_level: 2  |  |  | 0.8461 |
| quality_level: 3  |  |  | 0.8527 |
| quality_level: 4  |  |  | 0.8510 |

</details>

<details close>
<summary> ### SDXL </summary>

> compute_density_threshold = 100

- [sdxl:conv_linear_calibrate_info](https://ccssu.github.io/git_pages.io/run_sdxl/conv_linear_calibrate_infoUNet2DConditionModel_34f5bd9d3416cc1703c40491d1fa89fc50081c344f7a9ccf7b64f853a711f342.html)
- [sdxl:const_cailbrate_info](https://ccssu.github.io/git_pages.io/run_sdxl/costs_calibrate_info_UNet2DConditionModel_34f5bd9d3416cc1703c40491d1fa89fc50081c344f7a9ccf7b64f853a711f342.json.html)

| sdxl | quality_level | SSIM               | Conv-Count | Linear-Count |
| ---- | ------------- | ------------------ | ---------- | ------------ |
| sdxl | 0             | 0.9426063189586985 | 51         | 743          |
| sdxl | 1             | 0.9531199491728994 | 49         | 722          |
| sdxl | 2             | 0.9615541090957868 | 45         | 659          |
| sdxl | 3             | 0.960556455965684  | 42         | 687          |
| sdxl | 4             | 0.9626790998577764 | 34         | 500          |

|                  | PyTorch                                                  | OneDiff                                                    | SSIM   |
| ---------------- | -------------------------------------------------------- | ---------------------------------------------------------- | ------ |
|                  |  |  | 0.9889 |
| quality_level: 0 |  |        | 0.9426 |
| quality_level: 1 |  |        | 0.9531 |
| quality_level: 2 |  |        | 0.9616 |
| quality_level: 3 |  |        | 0.9606 |
| quality_level: 4 |  |        | 0.9627 |

</details>

<details close>
<summary> ### SDXL DeepCache </summary>

> compute_density_threshold = 100

| sdxl_deepcache | quality_level | SSIM               | Conv-Count | Linear-Count |
| -------------- | ------------- | ------------------ | ---------- | ------------ |
| sdxl_deepcache | 0             | 0.8990330836320674 | 0          | 743          |
| sdxl_deepcache | 1             | 0.9007828115010863 | 0          | 722          |
| sdxl_deepcache | 2             | 0.9045531115486659 | 0          | 653          |
| sdxl_deepcache | 3             | 0.9066207210614762 | 0          | 561          |
| sdxl_deepcache | 4             | 0.9074547844675099 | 0          | 504          |

| sdxl_deepcache | PyTorch  | OneDiff| SSIM | 
|---|---|---|---|
|  | |   | 0.9061| 
| quality_level: 0 | |   | 0.8990 |
| quality_level: 1 | |   | 0.9008 |
| quality_level: 2 | |   | 0.9046 |
| quality_level: 3 | |   | 0.9066 |
| quality_level: 4 | |   | 0.9075 |

</details>

<details>
<summary> demo.py </summary>

```python

"""
CUDA_VISIBLE_DEVICES=7 python sdxl.py --model_id /share_nfs/hf_models/stable-diffusion-xl-base-1.0 \
                                        --seed 1 \
                                        --backend onediff 

CUDA_VISIBLE_DEVICES=7 python sdxl.py --model_id /share_nfs/hf_models/stable-diffusion-xl-base-1.0 \
                                        --seed 1 \
                                        --quality_level 0 \
                                        --compute_density_threshold 0 \
                                        --backend onediff \
                                        --quantize \
                                        --cache_dir ./run_sdxl_quant 
"""
import os 
from diffusers import StableDiffusionXLPipeline
import torch
import argparse
from onediffx import compile_pipe
import time

class Time:
    def __enter__(self):
        self.start_time = time.time()
        return self 
    
    def __exit__(self, *args, **kwargs):
        self.end_time = time.time()
        self.duration = self.end_time - self.start_time
        print(f"Duration: {self.duration:.2f} seconds")
        return self
    

    
def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_id", type=str, default="/share_nfs/hf_models/stable-diffusion-xl-base-1.0 ")
    parser.add_argument("--prompt", type=str, default="a photo of an astronaut riding a horse on mars")
    parser.add_argument("--output_file", type=str, default="astronaut_rides_horse_onediff_quant.png")
    parser.add_argument("--seed", type=int, default=1)
    parser.add_argument("--quality_level", type=int, default=2, choices=list(range(8)))
    parser.add_argument("--compute_density_threshold", type=int, default=100)
    parser.add_argument("--backend", type=str, default="onediff", choices=["onediff", "torch"])
    parser.add_argument("--quantize", action="store_true")
    parser.add_argument("--cache_dir", type=str, default="./run_sdxl")
    args = parser.parse_args()
    return args
args = parse_args()
pipe = StableDiffusionXLPipeline.from_pretrained(args.model_id, torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
pipe.to("cuda")


def auto_pipe(pipe, cache_dir: str):
    compiled_options = {
        "options":{
            "graph_file": os.path.join(cache_dir, "tmp.graph"),
        }
    }
    pipe = compile_pipe(pipe, compiled_options=compiled_options)
    pipe.unet.quantize(**quantize_config)
    # "vae.decoder",
    # "vae.encoder",
    # pipe.vae.decoder = oneflow_compile(pipe.vae.decoder)
    # pipe.vae.decoder.quantize(**quantize_config)
    # pipe.vae.encoder = oneflow_compile(pipe.vae.encoder)
    # pipe.vae.encoder.quantize(**quantize_config)
    # "text_encoder",
    # pipe.text_encoder = oneflow_compile(pipe.text_encoder)
    pipe.text_encoder.quantize(**quantize_config)
    # "text_encoder_2",
    # pipe.text_encoder_2 = oneflow_compile(pipe.text_encoder_2)
    pipe.text_encoder_2.quantize(**quantize_config)
    # "image_encoder",
    # pipe.image_encoder = oneflow_compile(pipe.image_encoder)
    # pipe.image_encoder.quantize(**quantize_config)
    return pipe

    
prompt = "street style, detailed, raw photo, woman, face, shot on CineStill 800T"
graph_dir = os.path.join(args.cache_dir, "graph")

torch.manual_seed(args.seed)
if args.backend == "onediff":
    # cache_dir=args.cache_dir, quality_level=args.quality_level, compute_density_threshold=args.compute_density_threshold
    quantize_config = {
        "cache_dir": args.cache_dir,
        "quality_level": args.quality_level,
        "compute_density_threshold": args.compute_density_threshold,
        "quantize_conv": True,
    }
    if args.quantize:
        pipe = auto_pipe(pipe, graph_dir)


# warmup
for _ in range(6):
    with Time():
        images = pipe(prompt="a big dog", height = 1024, width = 1024, num_inference_steps=30).images[0]

with Time():
    torch.manual_seed(args.seed)
    images = pipe(prompt=prompt, height = 1024, width = 1024, num_inference_steps=30).images[0]

print(f'save to {args.output_file}')
images.save(args.output_file)


```

<details open> 
<summary> output </summary>

```shell
Duration: 3.15 seconds
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:02<00:00, 11.14it/s]
Duration: 3.16 seconds
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:02<00:00, 11.12it/s]
Duration: 3.16 seconds
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:02<00:00, 11.21it/s]
Duration: 3.13 seconds
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:02<00:00, 11.20it/s]
Duration: 3.14 seconds
save to astronaut_rides_horse_onediff_quant.png
```
</details>

</details>

### CODE REVIEW

File: src/onediff/infer_compiler/with_animatediff_compile/quantization_module.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+        calculator = create_quantization_calculator(
+            self.torch_model,
+            config,
+            basic_cache_key,
+            quality_level=config.quality_level,
+        )
+        out = calculator.calibrate(*args, **kwargs)
+        config.linear_count, config.conv_count = self.get_layer_counts(out.keys())
+        config.calibrate_info = out
+        config.save_quantization_config(file_name="calibrate_info.json")
```
Comment (ccssu) [inline src/onediff/infer_compiler/with_animatediff_compile/quantization_module.py:558]: 获取量化校准信息

File: src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py [hunk 2]
```diff
@@ -94,48 +105,10 @@ def to(self, *args, **kwargs):
     def _quantize_forward_faster(self, *args, **kwargs):
```
Comment (hjchen2) [inline src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py:105]: 我建议再抽象出一个OnlineQuantDeployableModule，quantize相关的接口都挪到那里去

File: src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py [hunk 3]
```diff
@@ -94,48 +105,10 @@ def to(self, *args, **kwargs):
...
-            ):
-                continue
-            if costs.get_compute_density(name) > compute_density_threshold:
-                calibrate_info[name] = "1"
-
-        self.quantization_config.calibrate_info = calibrate_info
+        # prepare and calibration
+        QuantizationModule(
+            self._deployed_original_module, self.quantization_config
+        ).quantize_forward(*args, **kwargs)
```
Comment (hjchen2) [inline src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py:111]: 最好可以复用QuantPipeline的代码，可以把具体量化的实现抽象出来，QuantPipeline和这里复用起来

Comment (hjchen2) [inline src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py:111]: 可以把具体量化的实现抽象出来，QuantPipeline和这里复用一套实现

Comment (ccssu) [reply src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py:111]: 开始是想直接用 onediff/src/onediff/quantization/quantize_pipeline.py 中 QuantPipeline， 但是**内部耦合 vae** ，不能复用。

File: src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py [hunk 4]
```diff
@@ -27,28 +30,36 @@ def quantize(
         quantize_conv=True,
         quantize_linear=True,
         bits=8,
-        compute_density_threshold=10,
+        quality_level=2,
```
[the remark this answers is not in the document]
Comment (ccssu) [reply src/onediff/infer_compiler/with_animatediff_compile/deployable_module.py:33]: quality_level 设置比较简单，数字越大，精度越高，加上 可调 compute_density_threshold 大小可以 更精确控制。
目前对 sdxl, sd1.5,sd2.1 ， 和组合 deepcache 提供 推荐设置 对用户上手简单，可以量化自己的模型。

精确控制量化结果如下代码：（实验我用了0～7后面发现 0～4已经够用了）
image
image

File: src/onediff/infer_compiler/with_animatediff_compile/quantization_module.py [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+        pipe = Pipe(self.model, args, kwargs)
+        config = self.config
+        quantizable_modules = find_quantizable_modules(
+            self.model, module_cls=[nn.Linear, nn.Conv2d]
+        )
+        costs_calibrate_info: Dict[str, Dict[str, float]] = config.load_calibrate_info(
+            self.costs_file_name
+        )
+        if len(costs_calibrate_info) == 0:
+            costs = metric_quantize_costs(pipe, {}, quantizable_modules)
```
Comment (ccssu) [inline src/onediff/infer_compiler/with_animatediff_compile/quantization_module.py:219]: todo: metric_quantize_costs :

**Pull request closed without merging**